core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java

```diff
@@ -0,0 +1,304 @@
+                final String writeMeterId_ipv6_addr =
+                        "armeria.client.dns.queries.written#count{name=foo.com.,server=0:0:0:0:0:0:0:1}";
+                final String timeoutMeterId =
+                        "armeria.client.dns.queries#count{" +
+                        "cause=resolver_timeout,name=foo.com.,result=failure}";
+                assertThat(MoreMeters.measureAll(meterRegistry))
+                        .doesNotContainKeys(writeMeterId_ipv4_addr,writeMeterId_ipv6_addr, timeoutMeterId);
+
+                assertThatThrownBy(() -> client2.execute(RequestHeaders.of(HttpMethod.GET, "http://foo.com"))
+                        .aggregate().join())
+                        .hasCauseInstanceOf(UnprocessedRequestException.class)
+                        .hasRootCauseExactlyInstanceOf(DnsTimeoutException.class);
+
+                await().untilAsserted(() -> {
+                    assertThat(MoreMeters.measureAll(meterRegistry))
+                            .containsAnyOf(entry(writeMeterId_ipv6_addr, 1.0),
+                                    entry(writeMeterId_ipv4_addr, 1.0));
+                });
+            }
+        }
+    }
+
+    @Test
+    void nxDomain() {
+        try (TestDnsServer server = new TestDnsServer(ImmutableMap.of(
+                new DefaultDnsQuestion("bar.com.", A),
+                new DefaultDnsResponse(0, DnsOpCode.QUERY, DnsResponseCode.NXDOMAIN)
+        ))) {
+            final MeterRegistry meterRegistry = PrometheusMeterRegistries.newRegistry();
+
+            try (ClientFactory factory = ClientFactory.builder()
+                    .domainNameResolverCustomizer(builder -> {
+                        builder.dnsServerAddressStreamProvider(dnsServerList(server));
+                        builder.searchDomains();
+                        builder.resolvedAddressTypes(ResolvedAddressTypes.IPV4_ONLY);
+                    })
+                    .meterRegistry(meterRegistry)
+                    .build()) {
+
+                final WebClient client = WebClient.builder()
+                        .factory(factory)
+                        .build();
+
+                final String writtenMeterId =
+                        "armeria.client.dns.queries.written#count{name=bar.com.,server=" +
+                        server.addr().getHostString() + '}';
+                final String nxDomainMeterId =
+                        "armeria.client.dns.queries#count{" +
+                        "cause=nx_domain,name=bar.com.,result=failure}";
+                assertThatThrownBy(() -> client.get("http://bar.com").aggregate().join())
+                        .hasRootCauseInstanceOf(UnknownHostException.class);
+
+                await().untilAsserted(() -> {
+                    assertThat(MoreMeters.measureAll(meterRegistry))
+                            .containsEntry(writtenMeterId, 2.0)
+                            .containsEntry(nxDomainMeterId, 2.0);
```

Comment:
Global comment: Could we also make sure that no other unexpected counters are created or increased for all test cases? e.g. the meter with `cause=others` should be 0 or non-existent.

Reply:
Gentle ping 😄

Reply:
I ran the test with some debug code to print the metric registry for each UT - 
The results below confirm we do not see cause=others any more.
----------------------------------------Running success----------------------------------------
{armeria.client.dns.queries.written#count{name=foo.com.,server=0:0:0:0:0:0:0:1}=1.0, armeria.client.dns.queries#count{cause=none,name=foo.com.,result=success}=1.0}
----------------------------------------Running timeout----------------------------------------
{armeria.client.dns.queries#count{cause=no_name_server_found,name=foo.com.,result=failure}=2.0, armeria.client.dns.queries#count{cause=resolver_timeout,name=foo.com.,result=failure}=1.0, armeria.client.dns.queries.written#count{name=foo.com.,server=0:0:0:0:0:0:0:1}=1.0}
----------------------------------------Running nxDomain----------------------------------------
{armeria.client.dns.queries#count{cause=nx_domain,name=bar.com.,result=failure}=2.0, armeria.client.dns.queries#count{cause=no_name_server_found,name=bar.com.,result=failure}=3.0, armeria.client.dns.queries.written#count{name=bar.com.,server=0:0:0:0:0:0:0:1}=2.0}
22:46:37.220 [Test worker] DEBUG c.l.a.i.c.DefaultDnsNameResolver - [bar.com] Sending a DNS query: DnsQuestion(bar.com IN A)
------------------------------------Running cname----------------------------------------
{armeria.client.dns.queries.cnamed#count{cname=baz.com.,name=bar.com.}=1.0, armeria.client.connections.lifespan#count{protocol=h2c}=0.0, armeria.client.connections.lifespan#total{protocol=h2c}=0.0, armeria.client.connections.lifespan#max{protocol=h2c}=0.0, armeria.client.dns.queries#count{cause=none,name=bar.com.,result=success}=2.0, armeria.client.dns.queries.written#count{name=bar.com.,server=0:0:0:0:0:0:0:1}=2.0}
----------------------------------------Running noAnswer----------------------------------------
com.linecorp.armeria.client.RefreshingAddressResolver$CacheEntry$$Lambda$595/0x0000000800f9f040@710be18c
{armeria.client.dns.queries.noanswer#count{code=10,name=bar.com.}=1.0, armeria.client.dns.queries#count{cause=nx_domain,name=bar.com.,result=failure}=1.0, armeria.client.dns.queries#count{cause=no_name_server_found,name=bar.com.,result=failure}=3.0, armeria.client.dns.queries.written#count{name=bar.com.,server=0:0:0:0:0:0:0:1}=2.0}

Reply:
Great, but could we automate the checks?

Reply:
Yes I have added them now.